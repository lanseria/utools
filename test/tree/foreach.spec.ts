import { treeForeach, ForeachType } from "@lanseria/utools";
import { data1, data2 } from "../mock";

describe("Tree Foreach 树循环", () => {
  describe("Dfs 深度遍历", () => {
    describe("Recursive 递归式", () => {
      it("List.length", () => {
        const list: INode[] = [];
        treeForeach(data2, (n) => {
          list.push(n);
        });
        expect(list.length).toBe(10);
        const nameList = list.map((m) => m.name);
        expect(nameList).toMatchSnapshot();
      });

      it("TreeList.length", () => {
        const treeList: INode[] = [];
        data1.map((m) => {
          treeForeach(m, (n) => {
            treeList.push(n);
          });
        });
        expect(treeList.length).toBe(20);
      });
    });

    describe("Non-Recursive 非递归式", () => {
      const treeForeachNonRecursive = (node: INode, cb: Fn) => {
        treeForeach(node, cb, ForeachType.DfsNonRecursive);
      };
      it("List.length", () => {
        const list: INode[] = [];
        treeForeachNonRecursive(data2, (n) => {
          list.push(n);
        });
        expect(list.length).toBe(10);
        const nameList = list.map((m) => m.name);
        expect(nameList).toMatchSnapshot();
      });

      it("TreeList.length", () => {
        const treeList: INode[] = [];
        data1.map((m) => {
          treeForeachNonRecursive(m, (n) => {
            treeList.push(n);
          });
        });
        expect(treeList.length).toBe(20);
      });
    });
  });
  describe("Bfs 广度遍历", () => {
    const treeForeachBfs = (node: INode, cb: Fn) => {
      treeForeach(node, cb, ForeachType.Bfs);
    };
    it("List.length", () => {
      const list: INode[] = [];
      treeForeachBfs(data2, (n) => {
        list.push(n);
      });
      expect(list.length).toBe(10);
      const nameList = list.map((m) => m.name);
      expect(nameList).toMatchSnapshot();
    });

    it("TreeList.length", () => {
      const treeList: INode[] = [];
      data1.map((m) => {
        treeForeachBfs(m, (n) => {
          treeList.push(n);
        });
      });
      expect(treeList.length).toBe(20);
    });
  });
});
