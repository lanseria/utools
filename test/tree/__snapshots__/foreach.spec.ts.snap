// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Foreach 树循环 Bfs 广度遍历 List.length 1`] = `
Array [
  "a",
  "a1",
  "a2",
  "a3",
  "a11",
  "a12",
  "a21",
  "a22",
  "a31",
  "a32",
]
`;

exports[`Tree Foreach 树循环 Dfs 深度遍历 Non-Recursive 非递归式 List.length 1`] = `
Array [
  "a",
  "a1",
  "a11",
  "a12",
  "a2",
  "a21",
  "a22",
  "a3",
  "a31",
  "a32",
]
`;

exports[`Tree Foreach 树循环 Dfs 深度遍历 Recursive 递归式 List.length 1`] = `
Array [
  "a",
  "a1",
  "a11",
  "a12",
  "a2",
  "a21",
  "a22",
  "a3",
  "a31",
  "a32",
]
`;
